### PR TITLE
[sonic-host-services]: Support GCU and reload

### PR DESCRIPTION
**Pull request by ganglyu**
Signed-off-by: Gang Lv [ganglv@microsoft.com](mailto:ganglv@microsoft.com)

Why I did it
GNMI needs to use host service to invoke generic_config_updater and config reload.

How I did it
Add host_modules for generic_config_updater and config reload.
Add unit test for host modules.

How to verify it
Run unit test for sonic-host-services.

### CODE REVIEW

**Comment by ganglyu** [comment]
azpw run sonic-net.sonic-host-services

**Comment by wen587** [comment]
GCU part LGTM.

File: host_modules/config_engine.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+        result = subprocess.run(cmd, shell=False, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        msg = ''
+        if result.returncode:
+            lines = result.stderr.decode().split('\n')
+            for line in lines[::-1]:
+                if 'Error' in line:
+                    msg = line
+                    break
+        return result.returncode, msg
```
Comment (liuh-80) [inline host_modules/config_engine.py:33]: Execute command and return error message code seems dupe in multiple places, suggest create a new method.

File: tests/host_modules/gcu_test.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    @mock.patch("dbus.SystemBus")
+    @mock.patch("dbus.service.BusName")
+    @mock.patch("dbus.service.Object.__init__")
+    def test_create_checkpoint(self, MockInit, MockBusName, MockSystemBus):
+        with mock.patch("subprocess.run") as mock_run:
+            res_mock = mock.Mock()
+            test_ret = 0
+            test_msg = b"Error: this is the test message\nHello world\n"
+            attrs = {"returncode": test_ret, "stderr": test_msg}
+            res_mock.configure_mock(**attrs)
```
Comment (liuh-80) [inline tests/host_modules/gcu_test.py:91]: Suggest create a new method for dupe code.

File: pytest.ini [hunk 3]
```diff
@@ -1,2 +1,2 @@
 [pytest]
-addopts = --cov=scripts --cov-report html --cov-report term --cov-report xml --ignore=tests/*/test*_vectors.py --junitxml=test-results.xml -vv
+addopts = --cov=scripts --cov=host_modules --cov-report html --cov-report term --cov-report xml --ignore=tests/*/test*_vectors.py --junitxml=test-results.xml -vv
```
Comment (qiluo-msft) [inline pytest.ini:2]: >host_modules [](http://example.com/codeflow?start=30&length=12)

Why need add this path? Is the default option working well?

Comment (ganglyu) [reply pytest.ini:2]: I have checked the coverage result, and the default setting does not cover host_modules directory.

**Review by qiluo-msft — APPROVED** [review]

File: pytest.ini (hunk 3 above)
Comment (qiluo-msft) [reply pytest.ini:2]: Could you just remove all the directories, like `--cov=scripts --cov=host_modules`?

Comment (ganglyu) [reply pytest.ini:2]: I have tried to remove "--cov=scripts --cov=host_modules", and pytest cannot generate coverage.xml.

**Comment by qiluo-msft** [comment]
@bhagatyj Could you or your team help review?

File: host_modules/config_engine.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    """
+    DBus endpoint that executes the config command
+    """
+    @host_service.method(host_service.bus_name(MOD_NAME), in_signature='s', out_signature='is')
+    def reload(self, config_db_json):
+
+        cmd = ['/usr/local/bin/config', 'reload', '-y']
+        config_db_json = config_db_json.strip()
+        if config_db_json and len(config_db_json):
+            config_file = '/tmp/config_db.json'
```
Comment (qiluo-msft) [inline host_modules/config_engine.py:20]: >config_file [](http://example.com/codeflow?start=12&length=11)

You can use the command `config load /dev/stdin -y`, no need to create a tmp file.

File: host_modules/gcu.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    @host_service.method(host_service.bus_name(MOD_NAME), in_signature='s', out_signature='is')
+    def apply_patch_db(self, patch_text):
+        patch_file_path = '/tmp/config_db.patch'
+        try:
+            with open(patch_file_path, 'w') as fp:
+                fp.write(patch_text)
+        except Exception as err:
+            return -1, "Fail to create patch file: %s"%str(err)
+
+        cmd = ['/usr/local/bin/config', 'apply-patch', '-f', 'CONFIGDB', patch_file_path]
```
Comment (qiluo-msft) [inline host_modules/gcu.py:21]: >patch_file_path [](http://example.com/codeflow?start=73&length=15)

also from stdin, so no need to create tmp file.

File: host_modules/gcu.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    @host_service.method(host_service.bus_name(MOD_NAME), in_signature='s', out_signature='is')
+    def apply_patch_yang(self, patch_text):
+        patch_file_path = '/tmp/config_yang.patch'
+        try:
+            with open(patch_file_path, 'w') as fp:
+                fp.write(patch_text)
+        except Exception as err:
+            return -1, "Fail to create patch file: %s"%str(err)
+
+        cmd = ['/usr/local/bin/config', 'apply-patch', '-f', 'SONICYANG', patch_file_path]
```
Comment (qiluo-msft) [inline host_modules/gcu.py:42]: >patch_file_path [](http://example.com/codeflow?start=74&length=15)

the same

File: host_modules/config_engine.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+class Config(host_service.HostModule):
+    """
+    DBus endpoint that executes the config command
+    """
+    @host_service.method(host_service.bus_name(MOD_NAME), in_signature='s', out_signature='is')
+    def reload(self, config_db_json):
+
+        cmd = ['/usr/local/bin/config', 'reload', '-y']
+        config_db_json = config_db_json.strip()
```
[the remark this answers is not in the document]
Comment (qiluo-msft) [reply host_modules/config_engine.py]: Still have the question.

[the remark this answers is not in the document]
Comment (ganglyu) [reply host_modules/config_engine.py]: > if config_db_json and len(config_db_json.strip()):
Now I use  len(config_db_json.strip()) to detect string with only spaces, and config_db_json is still the same.

**Review by qiluo-msft — APPROVED** [review]

**Pull request merged**